homeassistant/components/plugwise/sensor.py

```diff
@@ -0,0 +1,422 @@
+        """Icon for the sensor."""
+        if self._sensor_type is None:
+            if self._heating_state:
+                return FLAME_ICON
+            if self._cooling_state:
+                return COOL_ICON
+            return IDLE_ICON
+        return self._icon
+
+    def _process_data(self):
+        """Update the entity."""
+        data = self._api.get_device_data(self._dev_id)
+
+        if not data:
+            _LOGGER.error("Received no data for device %s.", self._name)
+        else:
+            if self._sensor in data:
+                if data[self._sensor] is not None:
+                    measurement = data[self._sensor]
+                    if self._sensor == "battery" or self._sensor == "valve_position":
+                        measurement = measurement * 100
+                    if self._unit_of_measurement == "%":
+                        measurement = int(measurement)
+                    self._state = measurement
+
+            if "heating_state" in data:
+                if data["heating_state"] is not None:
+                    self._heating_state = data["heating_state"]
+            if "cooling_state" in data:
+                if data["cooling_state"] is not None:
+                    self._cooling_state = data["cooling_state"]
+            if self._sensor == DEVICE_STATE:
+                if self._heating_state:
+                    self._state = "heating"
+                elif self._cooling_state:
+                    self._state = "cooling"
+                else:
+                    self._state = "idle"
+
+        self.async_write_ha_state()
+
+
+class PwPowerSensor(SmileGateway, Entity):
+    """Power sensor entities."""
+
+    def __init__(self, api, coordinator, name, dev_id, sensor, sensor_type, model):
+        """Set up the Plugwise API."""
+        super().__init__(api, coordinator)
+
+        self._api = api
+        self._model = model
+        self._name = name
+        self._dev_id = dev_id
+        self._device = sensor_type[0]
+        self._unit_of_measurement = sensor_type[1]
+        self._dev_class = sensor_type[2]
+        self._icon = sensor_type[3]
+        self._sensor = sensor
+        self._state = None
+
+        sensorname = sensor.replace("_", " ").title()
+        self._sensorname = f"{name} {sensorname}"
+
+        if self._dev_id == self._api.gateway_id:
+            self._name = f"Smile {self._name}"
+
+        self._unique_id = f"{dev_id}-{sensor}"
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._sensorname
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def device_class(self):
+        """Device class of this entity."""
+        return self._dev_class
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def device_info(self) -> Dict[str, any]:
```

Comment:
Looks like this is similar to climate. Can you make a base class that provides the device info for both?

Reply:
I added them to the baseclass that was there already and while at it created a new one from that that's shared between the two sensors. Way better like this.